options/options.go

```diff
@@ -122,20 +129,29 @@ func InterpretOptionsLenient(linked linker.Result, opts ...InterpreterOption) (s
 // effectively ignored. Any options that are uninterpretable (due to such
 // errors) will remain in the "uninterpreted_option" fields.
 func InterpretUnlinkedOptions(parsed parser.Result, opts ...InterpreterOption) (sourceinfo.OptionIndex, error) {
-	return interpretOptions(true, noResolveFile{parsed}, nil, reporter.NewHandler(nil), opts)
+	return interpretOptions(true, noResolveFile{parsed}, nil, nil, opts)
 }
 
 func interpretOptions(lenient bool, file file, res linker.Resolver, handler *reporter.Handler, interpOpts []InterpreterOption) (sourceinfo.OptionIndex, error) {
-	interp := interpreter{
+	interp := &interpreter{
 		file:       file,
 		resolver:   res,
 		lenient:    lenient,
-		reporter:   handler,
 		index:      sourceinfo.OptionIndex{},
 		pathBuffer: make([]int32, 0, 16),
 	}
+	var errHandler internal.ErrorHandler
+	switch {
+	case lenient:
+		errHandler = (*lenientHandler)(interp)
```

Comment:
The `handler` is ignored if lenient is enabled. Is that intended?

Reply:
As mentioned above, there is no `handler` when lenient is true. There are three entry points: one is non-lenient and requires a handler; the other two are lenient and do not accept a handler. This internal, unexported function accepts both values, but perhaps it would be more clear if I created a new type, like a `handlerPolicy`, with a lenient impl and a non-lenient one that requires a handler. I didn't do that because it felt like over-abstracting (and since it would have to use interfaces, there's a necessary allocation with that approach, too). I could also just try to clarify with a well-placed comment.

Reply:
This already feels over-abstracted. I think you could remove the `lenientHandler` type and check at the callsite where you would delegate to the error handler to avoid. It's difficult to understand that the `lenientHandler` here is just being used as an error constructor.

Reply:
On reading options.go I see that `interp.reporter.HandleErrorf` is used more often then I thought. To expand on my comment above. I think it would be more clear to have an internal method `interp.handleErrorf` that handle the lienent checks and then delegates to the underlying reporter. But non-blocking.

Reply:
The issue is that this code also uses a helper, `internal.FindFirstOption`, which takes a handler. So it needs to be a separate type in order to implement the interface that helper expects (which is a subset of the method set of `*reporter.Handler`)

Reply:
Though, actually, I think I see a way to simplify that. I could make that helper accept a function, not an interface, and could then implement your suggestion and pass it `interp.handleErrorf`.

Reply:
Okay, I pushed up a change that applies your suggestion. WDYT?

Reply:
LGTM